### PR TITLE
feat: implemented uno shout rule and organized game loop

### PR DESCRIPTION
**Pull request by caiovini980**
### CHANGES
- Implemented the possibility to shout UNO when with 2 cards on hand
- Give 2 cards to the player if not shouting UNO when needed
- Validation on the `InputManager`

### CODE REVIEW

File: UnoGame/Card/CardManager.h [hunk 1]
```diff
@@ -27,7 +27,7 @@ class CardManager : public ManagerBase
     std::stack<CardBehaviour> _tossDeck;
 
     int _initialDeckSize = 104;
-    int _initialHandSize = 7; // 7
+    int _initialHandSize = 3; // 7
```
Comment (caiovini980) [inline UnoGame/Card/CardManager.h:30]: Mus change this back to 7 cards, 3 was for test purposes only

Comment (psydack) [reply UnoGame/Card/CardManager.h:30]: Out of curiosity: If it was just for test, why did you commit?

File: UnoGame/Input/InputManager.cpp [hunk 2]
```diff
@@ -33,12 +33,25 @@ bool InputManager::IsValid(int choice, const GameStates& state)
...
 }
 
-bool InputManager::IsValid(int choice, PlayerBehaviour& player)
+bool InputManager::IsValid(int choice, PlayerBehaviour& player, bool canShoutUno) const
 {
     // accept from -1 to n - 1 (number of cards on hand - 1)
-    return (choice >= -1 && choice <= player.GetCards().size() - 1)
+    const int amountOfCardsOnPlayerHand = static_cast<int>(player.GetCards().size());
+
+    if (canShoutUno)
```
Comment (psydack) [inline UnoGame/Input/InputManager.cpp:41]: I believe you can simplify this in a single line if

```cpp
return 
    (choice >= 0 && choice <= amountOfCardsOnPlayerHand - 1) || 
    (choice == _inputForDrawCar) ||
    (canShoutUno && choice == _inputForDrawCard)
```

File: UnoGame/Input/InputManager.h [hunk 3]
```diff
@@ -10,12 +11,15 @@ class InputManager : public ManagerBase
...
+    bool IsValid(int choice, PlayerBehaviour& player, bool canShoutUno) const;
+    
     int GetMaxInputForMenu() const;
+    int GetDrawCardInput() const;
+    int GetShoutUNOInput() const;
 
 private:
     int _maxInputForMenu = 3;
     int _inputForDrawCard = -1;
+    int _inputForUnoShout = -2;
```
Comment (psydack) [inline UnoGame/Input/InputManager.h:23]: Are these const?

File: UnoGame/Card/CardManager.h (hunk 1 above)
Comment (caiovini980) [reply UnoGame/Card/CardManager.h:30]: Wasn't on purpose, I forgot to change it back to 7 before pushing the changes

File: UnoGame/Input/InputManager.cpp (hunk 2 above)
Comment (caiovini980) [reply UnoGame/Input/InputManager.cpp:41]: Nice tip!

File: UnoGame/Input/InputManager.h (hunk 3 above)
Comment (caiovini980) [reply UnoGame/Input/InputManager.h:23]: Yes, they must be consts, nice catch

Comment (caiovini980) [reply UnoGame/Input/InputManager.h:23]: I was trying to make them consts, but rider keep showing me this warning
image

Comment (caiovini980) [reply UnoGame/Input/InputManager.h:23]: It seems to be a bad habit to make member variables as consts. Private seems to be enough
image

[Reference](https://lesleylai.info/en/const-and-reference-member-variables/)

**Pull request merged**